### PR TITLE
Port some builtin functions

### PR DESCRIPTION
**Pull request by blmarket**
Trying some initial port before working more on it, so it's the best time to talk how to port remaining functions.

C code link:
- [nox_script_getWall_511EB0](https://github.com/noxworld-dev/opennox/blob/dev/src/server__script__builtin.c#L114)
- [nox_script_canInteract_513E80](https://github.com/noxworld-dev/opennox/blob/dev/src/server__script__builtin.c#L1425)

Also added `compare` function to evaluate both versions and compare the results. Later we can deprecate compare (after some testing?) and C builtins.

## Required sign-off

- [x] I confirm that my PR does not contain any commercial or protected assets and/or source code.
- [x] I agree in advance that my codes will be licensed automatically under the Apache License or similar BSD/MIT-like
      open source licenses in case if OpenNox Project will adopt such a non-GPL license in the future.

### CODE REVIEW

File: src/script_noxscript_builtins.go [hunk 1]
```diff
@@ -1045,3 +1047,61 @@ func nox_script_GetScore_516EA0() int {
...
+	if v2 != nil && v3 != nil {
+		p2 := v2.Pos()
+		p3 := v3.Pos()
+		if abs(p2.X-p3.X) <= 512.0 && abs(p2.Y-p3.Y) <= 512.0 {
+			if nox_xxx_unitCanInteractWith_5370E0(v2, v3, 0) {
+				v4 = 1
+			}
+		}
+	}
+	noxServer.noxScript.PushI32(int32(v4))
```
Comment (dennwc) [inline src/script_noxscript_builtins.go]:
```suggestion
	s.PushI32(int32(v4))
```

File: src/script_noxscript_builtins.go [hunk 2]
```diff
@@ -1045,3 +1047,61 @@ func nox_script_GetScore_516EA0() int {
...
+
+func nox_script_getWall_511EB0() int {
+	s := &noxServer.noxScript
+
+	y := s.PopI32()
+	x := s.PopI32()
+
+	if s.f40 != 0 || s.f44 != 0 {
+		x = (int32(s.f40) + 23*x) / 23
+		y = (int32(s.f44) + 23*y) / 23
```
Comment (dennwc) [inline src/script_noxscript_builtins.go]:
```suggestion
		x = (int32(s.f40) + common.GridStep*x) / common.GridStep
		y = (int32(s.f44) + common.GridStep*y) / common.GridStep
```

File: src/script_noxscript_builtins.go [hunk 3]
```diff
@@ -1045,3 +1047,61 @@ func nox_script_GetScore_516EA0() int {
 	s.PushI32(int32(pl.lessons))
 	return 0
 }
+
+func compare(name string, orig func() int, updated func() int) func() int {
```
Comment (dennwc) [inline src/script_noxscript_builtins.go]:
```suggestion
func noxScriptCompare(name string, orig func() int, updated func() int) func() int {
```

Comment (dennwc) [inline src/script_noxscript_builtins.go]: Cool idea! Checking the stack for every builtin call would be expensive, though, so maybe we can only enable it when the debug mode is set? Or maybe compare stack down only to ~5 values max (checking max number of arguments + returns in all builtins + 1)? Also worth replacing `DeepEqual` with something more specialized (e.g. generic array compare function). In any case, I'm okay with merging it as-is and improving it later.

File: src/script_noxscript_builtins.go [hunk 4]
```diff
@@ -1045,3 +1047,61 @@ func nox_script_GetScore_516EA0() int {
...
+	return func() int {
+		s := &noxServer.noxScript
+		origStack := append([]uint32{}, s.vm.stack...)
+		ret1 := updated()
+		afterStack := append([]uint32{}, s.vm.stack...)
+		s.vm.stack = origStack
+		ret2 := orig()
+
+		if ret1 != ret2 || !reflect.DeepEqual(s.vm.stack, afterStack) {
+			log.Printf("%s: return or stack inequal %d vs %d\n", name, ret1, ret2)
```
Comment (dennwc) [inline src/script_noxscript_builtins.go]:
```suggestion
			scriptLog.Printf("%s: return or stack inequal %d vs %d\n", name, ret1, ret2)
```

File: src/script_noxscript_builtins.go (hunk 3 above)
Comment (blmarket) [reply src/script_noxscript_builtins.go]: Agree about its expensive computation cost, but hope we can revisit when the performance becomes an issue. Alternatively we can apply this to builtins not called frequently. (assuming frequently called builtins are relatively simple so not necessary to validate its logic)

re: slicing stack to smaller pieces: In some cases slicing the list could be more expensive than whole clone, as the logic gets more complicated and less cache friendly. Also we can reduce use of this utility if the performance matters.

Will replace DeepEqual with more appropriate func.

**Review by dennwc — APPROVED** [review]

**Pull request merged**